core/src/main/java/io/grpc/transport/AbstractStream.java

```diff
@@ -229,6 +278,74 @@ protected final void requestMessagesFromDeframer(int numMessages) {
     }
   }
 
+  /**
+   * Event handler to be called by the subclass when a number of bytes are being queued for sending
+   * to the remote endpoint. May call back the listener's {@link StreamListener#onReady()}
+   * handler if appropriate.
+   *
+   * @param numBytes the number of bytes being sent.
+   */
+  protected final void onSendingBytes(int numBytes) {
+    boolean shouldNotify = false;
+    synchronized (onReadyLock) {
+      shouldNotifyOnReady = true;
+      numSentBytesQueued += numBytes;
+      shouldNotify = shouldNotifyOnReady();
+    }
+    if (shouldNotify) {
+      listener().onReady();
```

Comment:
For efficiency of applications that obey outbound flow control, I think that immediately calling back during a write (and requiring the application code to be reentrant) is probably the best approach, so long as the documentation is clear. WDYT?